src/resource.rs

```diff
@@ -0,0 +1,143 @@
+use crate::borrow::{AtomicRefCell, Exclusive, Ref, RefMut, Shared};
+use std::{
+    any::{Any, TypeId},
+    collections::HashMap,
+    marker::PhantomData,
+    ops::{Deref, DerefMut},
+};
+
+pub enum ResourceAccessType {
+    Read,
+    Write,
+}
+
+pub trait Resource: 'static + Any + Send + Sync {}
+impl<T> Resource for T where T: 'static + Any + Send + Sync {}
+
+pub struct Read<'a, T: Resource> {
+    inner: Ref<'a, Shared<'a>, &'a T>,
+}
+impl<'a, T: Resource> Deref for Read<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target { self.inner.deref() }
+}
+
+/// TODO: RefMut::map_into seems to have borrowing issues when used in conjunction with this trait
+/// Hashmap. For now, I just pass the ref directly out, and perform the downcasting on Deref. This
+/// is not optimal, but it is a hack for this code to work for now until someone smarter can fix.
```

Comment:
I am not all that happy with the borrow module as a while. It's requirements have changed a few times and it is left more complex than it currently needs to be.